src/uwtools/drivers/forecast.py

```diff
@@ -0,0 +1,250 @@
+
+    run()
+        Runs the forecast executable with the namelist file and staged
+        input files. This will both build the executable and run it.
+    
+    '''
+
+    def __init__(self, argv):
+
+        '''
+        Initialize the Forecast driver.
+
+        '''
+        self.args = parse_args(argv)
+
+    def requirements(self):
+
+        ''' Recursively parse config and platform files to determine and
+         fill in any dependencies. '''
+
+    def resources(self):
+
+        ''' Determine necessary task objects and fill in resource requirements of each.
+         Returns a Config object containing the HPC resources needed.'''
+
+    def validate(self):
+
+        ''' Validates the objects generated by the driver from the provided
+        config and platform files. If the --dry_run flag is provided, complete all
+        stages through validation and print results without running the forecast.'''
+
+    def main(self, argv):
+        '''
+        Defines the user interface for the forecast driver. Parses arguments provided by the user
+        and passes to the Forecast driver class to be run.'''
+        self.args = parse_args(argv)
+
+    @abc.abstractmethod
+    def create_model_config(self):
+
+        ''' Collects all the user inputs required to create a model config 
+        file, calling the existing model config tools. This will be unique
+        to the app being run and will appropriately parse subsequent stages
+        of the workflow. 
+        Defaults will be filled in if not provided by the user. Equivalent 
+        references to config_default.yaml or config.community.yaml from SRW
+        will need to be made for the other apps. '''
+
+    @abc.abstractmethod
+    def stage_fix_files(self):
+
+        ''' Holds the knowledge for how to modify a list of fix files and 
+        stages them in the working directory. Likely gets all its info from 
+        config_obj. Calls data mover tool (could be python copy). Fix files 
+        usually are specific to a given named grid and resolution. '''
+
+    @abc.abstractmethod
+    def create_namelist(self):
+
+        ''' Collects all the user inputs required to create a namelist 
+        file, calling the existing namelist config tools. '''
+
+    @abc.abstractmethod
+    def stage_output_files(self):
+
+        ''' Holds the knowledge for how to modify a list of output files and 
+        stages them in the working directory. Output files usually are 
+        specific to a given app.'''
+
+    @abc.abstractmethod
+    def job_card(self):
+
+        ''' Turns the resources config object into a batch card for the 
+        configured Task. Can be called from the command line on the front end 
+        node to show the user what the job card would have looked like.'''
+
+    @abc.abstractmethod
+    def run(self):
+
+        ''' Runs the forecast executable with the namelist file and staged
+        input files. This will both build the executable and run it. '''
```

Comment:
The `@abc.abstractmethod` decorated methods here don't seem to be defined as I would have expected.

The ones that ALL drivers should conform to the same interface should be abstract methods, while the ones that are specific to a given component should be defined in that component's sub class, and may not necessarily have a strict interface. Those that have the same implementation in all Drivers can be defined fully in the base class (`run()` and `job_card()` _might_ be examples of those, but perhaps we don't want to assume that here).

Of the ones you've included, the ones that belong to all drivers include:

```
requirements
resources
validate
job_card
run
```

This set seems to be missing the `output()` method mentioned in the [docs](https://github.com/ufs-community/workflow-tools/wiki/Migrating-Production-Workflows-to-the-Unified-Workflow-using-the-Strangler-Fig-Pattern#component-drivers).

All the others should likely go in the subclass `FV3Forecast(Driver)` definition.

Reply:
So I kept `output()` as `stage_output_files()`, but I can change that. I think the methods otherwise are an issue due to the comments above, just a fundamental misunderstanding of the intended structure. Definitely overly simplistic and missing the necessary levels - thanks for the clarification, working on it now!

Reply:
The names of the two methods to me have significantly different meanings. The `stage_output_files()` method is a verb, and indicates that it's going to do some action, hopefully putting some files in a location. Whereas `output()` is a noun that says it will likely give me some _thing_, hopefully a data structure or object containing the expected "output" of the driver (maybe it's an action, files, a boolean value, or nothing at all).

Reply:
Ah, its more of a 'emphasis of priority' difference then! I was emphasizing the staging more than the return, but you're right that the actual output is more important than the directories and should be the focus - gotcha!